network/gossip/libp2p/peerManager.go

```diff
@@ -0,0 +1,165 @@
+	// return an error or create multiple connections
+	ConnectPeers(ctx context.Context, ids flow.IdentityList) map[flow.Identity]error
+
+	// DisconnectPeers disconnect from the given flow.Identities and returns a map of identities which failed.
+	// DisconnectPeers implementation should be idempotent such that multiple calls to connect to the same peer should
+	// not return an error
+	DisconnectPeers(ctx context.Context, ids flow.IdentityList) map[flow.Identity]error
+}
+
+// PeerUpdateInterval is how long the peer manager waits in between attempts to update peer connections
+var PeerUpdateInterval = 1 * time.Minute
+
+// PeerManager adds and removes connections to peers periodically and on request
+type PeerManager struct {
+	ctx         context.Context
+	logger      zerolog.Logger
+	idsProvider func() (flow.IdentityList, error) // callback to retrieve list of peers to connect to
+
+	peerRequestQ chan struct{}     // a channel to queue a peer update request
+	connector    Connector         // connector to connect or disconnect from peers
+	currentIDs   flow.IdentityList // a list of currentIDs this node should be connected to
+}
+
+// NewPeerManager creates a new peer manager which calls the idsProvider callback to get a list of peers to connect to
+// and it uses the connector to actually connect or disconnect from peers.
+func NewPeerManager(ctx context.Context, logger zerolog.Logger, idsProvider func() (flow.IdentityList, error), connector Connector) *PeerManager {
+	return &PeerManager{
+		ctx:          ctx,
+		logger:       logger,
+		idsProvider:  idsProvider,
+		connector:    connector,
+		peerRequestQ: make(chan struct{}, 1),
+	}
+}
+
+// Start kicks off the ambient periodic connection updates
+func (pm *PeerManager) Start() error {
+	go pm.updateLoop()
+	go pm.periodicUpdate()
+	return nil
+}
+
+// updateLoop triggers an update peer request when it has been requested
+func (pm *PeerManager) updateLoop() {
+	for {
+		select {
+		case <-pm.ctx.Done():
+			return
+		case <-pm.peerRequestQ:
+			pm.updatePeers()
+		}
+	}
+}
+
+// updateLoop request periodic connection update
+func (pm *PeerManager) periodicUpdate() {
+
+	// request initial discovery
+	pm.RequestPeerUpdate()
+
+	ticker := time.NewTicker(PeerUpdateInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			pm.RequestPeerUpdate()
+		case <-pm.ctx.Done():
+			return
+		}
+	}
+}
+
+// RequestPeerUpdate initiates an update to peer connections of this node
+func (pm *PeerManager) RequestPeerUpdate() {
+	select {
+	case pm.peerRequestQ <- struct{}{}:
+	default:
+	}
+}
+
+// updatePeers updates the peers by connecting to all the nodes provided by the idsProvider callback and disconnecting from
+// previous nodes that are no longer in the new list of nodes.
+func (pm *PeerManager) updatePeers() {
+
+	idProvider := pm.idsProvider
+	if idProvider == nil {
+		pm.logger.Error().Msg("failed to update peers: id provider callback is not set")
```

Comment:
I would do this check in the constructor, and throw an error at that point instead. 

If you were to keep it here, you'd probably want to `return` at this point

Reply:
i am removing that check all together